### PR TITLE
Implied Extensions

### PR DESCRIPTION
**Pull request by cmoesel**
**OVERVIEW**

This PR adds support for "implied" extensions.  The basic premise of implied extensions is that you can use a particular extension URL format to reference resource elements from past or future versions of FHIR.  For example, the following extension references the STU3 `Patient.animal` element:
```
http://hl7.org/fhir/3.0/StructureDefinition/extension-Patient.animal
```
Since `Patient.animal` is actually a `BackboneElement` with three child elements (`species`, `breed`, and `genderStatus`), this extension is complex and has sub-extensions corresponding to those child elements.  If we references `Patient.animal.species` directly, however, it would be a simple extension w/ `value[x]` constrained to `CodeableConcept` and bound to the proper species value set.

Other behaviors of implied extensions:
* If the element is a single type and that type does not exist in your IG's version of FHIR, it will represent the extension as a complex extension w/ sub-extensions matching the elements of the unsupported type.  For example, an R4 IG that references an implied extension from R5 whose type is `CodeableReference` would end up being a complex extension with `concept` and `reference` sub-extensions.
* If the element references any types that have been renamed, then it will convert the type to the new name.  For example, an R4 IG that references an implied extension from DSTU2 whose type is `Reference(MedicationOrder)` would be converted to an extension w/ `value[x]` type `Reference(MedicationRequest)`.
* If the element is a choice and some of the choices do not exist in the IG's version of FHIR (and also don't have corresponding renamed types), then those types will be dropped from the types of the resulting extension.

Support for this feature required a mechanism for loading "supplemental" versions of FHIR.  I extended the `FHIRDefinitions` class to support holding these supplemental versions of FHIR -- and the `FHIRDefinitions.fishForFHIR` function was updated to recognize the special URL format and "materialize" these implied extensions if the corresponding supplemental version of FHIR is loaded.

Since we don't want to load every version of FHIR for every run of SUSHI, and since implied extensions are not likely to be widely used, we require users to indicate that they will use implied extensions by specifying the corresponding extensions package as a dependency. These are, in fact, "virtual" dependency packages because they do not actually exists (despite being listed in the FHIR documentation).  Valid virtual extension packages are:
* hl7.fhir.extensions.r2
* hl7.fhir.extensions.r3
* hl7.fhir.extensions.r4
* hl7.fhir.extensions.r5

When specifying them as a dependency, their specified version should match the same version of FHIR as your IG uses.  For example, if your IG uses FHIR R4 but wants to reference an implied extension for R3, it would specify the following in `sushi-config.yaml`:
```
dependencies:
  hl7.fhir.extensions.r3: 4.0.1
```

**BONUS FEATURES**

This PR has a few bonus things you may notice in the code.  Most notably:
* Some dependency updates that were required to resolve strange bugs in Jest
* Refactoring mixins to a separate file to resolve circular import issues
* Refactoring some of the loading functions to have more sensible return types
* Refactoring some of the tests to separate R5 tests into their own suite (which was necessary, as my addition of many new files in `r5-definitions` caused some existing tests to break since they were mixing R4 and R5 together).

**TESTING**

There are unit tests for the different combinations of things and for the known edge cases.  Feel free to check them out.  Also feel free to try the attached FSH project that has some illustrative FSH definitions in it already.  Note that we actually support this feature _better_ than the IG Publisher does, so you'll notice that if you build it in the IG Publisher, some things will not be as pretty as you expect and some of the more advanced features will cause (incorrect) errors in the IG Publisher.

[ImpliedExtensionsProject.zip](https://github.com/FHIR/sushi/files/6355340/ImpliedExtensionsProject.zip)

**MORE INFO**

FHIR Documentation for this feature is here: http://build.fhir.org/versions.html#extensions

A thrilling Zulip conversation is here: https://chat.fhir.org/#narrow/stream/215610-shorthand/topic/Use.205.2E0.20Extension/near/226875574

### CODE REVIEW

**Comment by cmoesel** [comment]
How embarrassing.  But honestly, these tests work on my system!  Can someone else give them a try?  Having another data point might help me try to determine how to attack this.

**Comment by cmoesel** [comment]
OK.  I found and fixed the issue w/ the failing tests.

The `TestFisher` loads resources from your local FHIR cache when it can't find them in the test package.  When it does this, it copies them over to the test package.  Prior to my fix in aa9aeaa, it _always_ loaded from the FHIR R4 cache and always stored loaded resources back to the (mostly FHIR R4) package.

The problem was that I refactored the R5-based tests to use only the R5 test package (`r5-definitions`).  So when the `TestFisher` couldn't find a resource in the R5 test package, it loaded the resource from the R4 cache and used it instead.  On our local systems, this worked well enough and there was no user-facing indication that it was happening -- so we didn't notice.  And since it copied the loaded resource into the R4-based test package, there was no diff either (and it didn't solve the underlying issue of the resource not being available in the R5-based test package).  In CI, however, there is no local FHIR cache, so the `TestFisher` fails to find the resource entirely and tests fail as a result.

I fixed this by updating the `TestFisher` so it can optionally take other cache package names and test package names -- so we can specify to load from the R5 cache and store in the R5 test package.  I also added a loud `console.log` statement to make it more obvious when things are loaded from the FHIR cache (since that is an exceptional situation and the user may want to be aware).

File: src/utils/Processing.ts [hunk 1]
```diff
@@ -203,19 +213,43 @@ export function loadExternalDependencies(
...
-        })
-    );
-  }
-  return dependencyDefs;
+  });
+
+  return Promise.all(promises).then(() => {});
+}
+
+export async function loadSupplementalFHIRPackage(
```
Comment (ngfreiter) [inline src/utils/Processing.ts:242]: Should this go in `load.ts`? I thought `loadExternalDependencies` only gets its special spot in `Processing.ts` because it is used directly in `app.ts`.

File: src/utils/Processing.ts [hunk 2]
```diff
@@ -203,19 +213,43 @@ export function loadExternalDependencies(
...
+      if (dep.version !== fhirVersion) {
+        logger.warn(
+          `Incorrect package version: ${dep.packageId}#${dep.version}. FHIR extensions packages ` +
+            "should use the same version as the implementation guide's fhirVersion. Version " +
+            `${fhirVersion} will be used instead. Update the dependency version in ` +
+            'sushi-config.yaml to eliminate this warning.'
+        );
+      }
+      logger.info(
+        `Loading supplemental version of FHIR to support virtual extensions package: ${dep.packageId}`
```
Comment (ngfreiter) [inline src/utils/Processing.ts:229]: I think the word `virtual` is a bit confusing. Would we be better off explaining a bit more that you are `Loading supplemental version of FHIR to support extensions from other versions of FHIR required by package: ${dep.packageId}`. Or something along those lines.

**Review by ngfreiter — COMMENTED** [review]
I've still got much to look at, but I don't trust GitHub entirely to save my comments if I leave them overnight, so here are some minor comments.

File: src/utils/Processing.ts (hunk 2 above)
Comment (ngfreiter) [reply src/utils/Processing.ts:229]: I'm not sure I like my message either to be honest, I just know the word virtual confused me, and I feel like most people using SUSHI won't care (and don't really need to care, assuming all goes well) that the extensions are "virtual".

File: src/app.ts [hunk 3]
```diff
@@ -143,7 +143,7 @@ async function app() {
 
   // Load dependencies
   const defs = new FHIRDefinitions();
-  const dependencyDefs = loadExternalDependencies(defs, config);
+  await loadExternalDependencies(defs, config);
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/app.ts:146]: TBH, I don't remember for sure.  I think it probably was because I changed the signature to return `Promise<void>` and I couldn't think of what to call the variable in that case -- and then I realized we don't do anything with it but just `await` it a few lines down.  So I probably "streamlined" it by making 2 lines into one and eliminating the need to figure out the name of the variable.

I don't know how much of a performance difference it really makes in this case, but I can move the await back down below `loadCustomResources`.

File: src/utils/Processing.ts (hunk 2 above)
Comment (cmoesel) [reply src/utils/Processing.ts:229]: Yeah.  I don't love the phrase "implied extensions" and "virtual extensions package" is also a little problematic. It would be easier if the spec gave us some real names for these ideas, but it doesn't.  But I want to say *something*, otherwise the user will likely be confused as to why we are downloading the core FHIR DSTU2 package (or advanced users might wonder why the requested extensions package isn't showing up in their FHIR cache).

Comment (cmoesel) [reply src/utils/Processing.ts:229]: I think your suggested message is close.  How about this? `Loading supplemental version of FHIR to support extensions from ${dep.packageId}`

File: src/utils/Processing.ts (hunk 1 above)
Comment (cmoesel) [reply src/utils/Processing.ts:242]: I think I put it here originally because it was initially a `private` function (used only by `loadExternalDependencies`).  But then I had to make it `public` in order to unit test it independently.

I can move it.

File: src/app.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ngfreiter) [reply src/app.ts:146]: Yeah I don't think the performance difference would really be considerable, I was just wondering why the change. I leave it up to you whether you think the simplicity of the new code is worth the maybe slight performance difference.

File: src/utils/Processing.ts (hunk 2 above)
Comment (ngfreiter) [reply src/utils/Processing.ts:229]: Yep, that message looks good!

File: src/fhirdefs/impliedExtensions.ts [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+  const rootElement = ext.findElement('Extension');
+  applyMetadataToElement(rootElement, ed);
+  applyContent(sd, ed, version, ext, rootElement, [], defs, supplementalDefs);
+
+  return ext.toJSON(true);
+}
+
+function isUnsupported(ed: any): boolean {
+  // According to documentation, elements with type "Resource" are not supported
+  return ed.type?.some((t: any) => t.code === 'Resource');
```
Comment (ngfreiter) [inline src/fhirdefs/impliedExtensions.ts:127]: Again getting theoretical here because I'm not sure if this case exists in the spec, but what if the type was something like `Patient`? Or another child of `Resource`?

**Review by ngfreiter — COMMENTED** [review]
Would you be able to add summary comments to the functions in `ImpliedExtensions`? I know the names themselves are descriptive, but for me I just can't entirely track what we are doing, and I think some summaries would go a long way for future generations.

File: src/fhirdefs/impliedExtensions.ts [hunk 5]
```diff
@@ -0,0 +1,457 @@
+// This file contains logic to support "implied" extensions, described in the FHIR spec here:
+// http://hl7.org/fhir/versions.html#extensions
+
+import { union } from 'lodash';
+import { logger, Type } from '../utils';
+import { ElementDefinition, ElementDefinitionType, StructureDefinition } from '../fhirtypes';
+import { FHIRDefinitions } from '../fhirdefs';
+
+export const IMPLIED_EXTENSION_REGEX = /^http:\/\/hl7\.org\/fhir\/([1345]\.0)\/StructureDefinition\/extension-(([^./]+)\.[^\/]+)$/;
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts:9]: > I'm trying to figure out why in (([^./]+)\.[^\/]+), the first / does not need to be escaped. But it seems like it doesn't...but it still scares me since ole reliable https://regex101.com/ does want it to be escaped.

It does not have to be escaped because it is a literal character inside the `[` `]` selector.  I think the only characters you need to escape inside those fancy square brackets are `\` and `]`.  When you try it in regex101.com, make sure you pick the ECMAScript flavor.  I suspect you were looking at it using the PCRE2 flavor (which _does_ require the `/` to be escaped).

[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts:9]: > I'm wondering why you disallow `/` in the part after the first `.`.

Yeah, I'm wondering that too.  It's probably not necessary.  That said, it also isn't really harmful because FHIR resources and types don't slice (and certainty don't reslice).

But still, I reckon we can get away with one of them `.*` regex black holes here.

File: src/fhirdefs/impliedExtensions.ts [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+
+export function isImpliedExtension(url: string) {
+  return IMPLIED_EXTENSION_REGEX.test(url);
+}
+
+export function materializeImpliedExtension(url: string, defs: FHIRDefinitions): any {
+  const match = url.match(IMPLIED_EXTENSION_REGEX);
+  if (match == null) {
+    logger.error(
+      `Cannot materialize implied extension (${url}) since the URL does not match the implied extension pattern ` +
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts:76]: You could at least do me the courtesy of copying my bad error message correctly. ;-)

So is your issue w/ the fact that we expose the idea of "materializing" extensions to the user?  Would you prefer something more like:

> Unsupported extension URL: ${url}.  URLs for converting between versions of FHIR must follow the pattern http://hl7.org/fhir/[version]/StructureDefinition/extension-[Path].  See: http://hl7.org/fhir/versions.html#extensions.

File: src/fhirdefs/impliedExtensions.ts [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+
+  const [, version, id, type] = match;
+  const supplementalPackage = VERSION_TO_PACKAGE_MAP[version]; // guaranteed not to be null thanks to REGEX check above
+  const supplementalDefs = defs.getSupplementalFHIRDefinitions(supplementalPackage);
+  if (supplementalDefs == null) {
+    const extRelease = version === '1.0' ? 'r2' : `r${version[0]}`;
+    const fhirVersion = defs.fishForFHIR('StructureDefinition', Type.Resource)?.fhirVersion;
+    logger.error(
+      `Cannot materialize implied extension: ${url}.\n` +
+        'To fix this, add the following dependency to your sushi-config.yaml file:\n' +
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts:90]: How about:
> The extension ${url} requires the following dependency to be declared in your sushi-config.yaml file:

File: src/fhirdefs/impliedExtensions.ts [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+        `  hl7.fhir.extensions.${extRelease}: ${fhirVersion}`
+    );
+    return;
+  }
+
+  const sd = supplementalDefs.fishForFHIR(type);
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts:96]: It hurts my fingers.  But only a tiny tiny bit.  I actually don't even notice most of the time.  I will try to muddle through it.

File: src/fhirdefs/impliedExtensions.ts [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+  if (sd == null) {
+    logger.error(
+      `Cannot materialize implied extension (${url}) since ${type} is not a valid resource or data type in ${supplementalPackage}`
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts:99]: We're not consistent about it at all. (I just checked). I thought that I was adding a period at the end if the error was multiple sentences, but leaving it off if it was just a phrase. But it seems that is not actually what I was doing. Oops. I can add periods.

And by the way, I appreciate the nitpickiness. I usually try to avoid inconsistencies like this -- so appreciate your help in pointing them out.

File: src/fhirdefs/impliedExtensions.ts (hunk 4 above)
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts:127]: This is a direct implementation of a prohibition from the spec:
> There is no way to represent the few special elements that have the type "Resource" in extensions, so these are not defined (e.g. Bundle.entry.resource, DomainResource.contained, etc)

But I think the main issue is that `Resource` is abstract.  There is no way to create an implied extension that could work for any possible value of it.  Other abstract resources are not surfaced as element types, so we don't need to worry about them.  And concrete resources (like `Patient`) _can_ be represented in extensions just fine.  If it's a _reference_ to a concrete resource, though, then that can only be represented if the resource (or its equivalent) exists in the version of FHIR you're bringing it into (since we can't represent a _reference_ to a complex extension).

**Comment by cmoesel** [comment]
Thanks for the review, @ngfreiter!  I've responded to some of your comments.  I'll address the issues when I get back next week!

File: src/fhirdefs/impliedExtensions.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ngfreiter) [reply src/fhirdefs/impliedExtensions.ts:76]: Ah yes, I do like that message.

File: src/fhirdefs/impliedExtensions.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ngfreiter) [reply src/fhirdefs/impliedExtensions.ts:90]: Sounds good!

**Review by mint-thompson — DISMISSED** [review]
This is a big one, but I think I've got a decent handle on it thanks to some discussions with @cmoesel . I can't think of any recommendations for things to change, so, approved!

File: src/utils/Processing.ts (hunk 1 above)
Comment (cmoesel) [reply src/utils/Processing.ts]: Fixed in e4448ef

File: src/utils/Processing.ts (hunk 2 above)
Comment (cmoesel) [reply src/utils/Processing.ts]: Fixed in ce58754

File: src/fhirdefs/impliedExtensions.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts]: Fixed in ce58754

File: src/fhirdefs/impliedExtensions.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts]: Fixed in ce58754

File: src/fhirdefs/impliedExtensions.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts]: Fixed in ce58754

File: src/fhirdefs/impliedExtensions.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts]: Fixed in ce58754

File: src/fhirdefs/impliedExtensions.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/fhirdefs/impliedExtensions.ts]: Fixed in ce58754

**Comment by cmoesel** [comment]
OK @mint-thompson and @ngfreiter -- this should be truly all set now.  Merged, circular dependencies _gone_, and feedback addressed.

**Review by ngfreiter — APPROVED** [review]
Thanks for adding the comments, such a tricky task, but so well done, looks good to me!

**Review by mint-thompson — APPROVED** [review]
